src/Form/Control/TreeItemSelector.php

```diff
@@ -102,6 +106,36 @@ public function onItem(\Closure $fx): void
         }, ['data' => 'value']);
     }
 
+    /**
+     * @param mixed $parentId
+     */
+    protected function addNodes(Model $model, $parentId = null): array
+    {
+        $result = [];
+
+        $nodeModel = (clone $model)->addCondition($this->parentIdField, $parentId);
+
+        foreach ($nodeModel as $node) {
```

Comment:
This imply a DB query (Model::getIterator() which implies SELECT). Recursively.

Single query recursive data fetch can be hard to implement and will require RECURSIVE CTE.

A good enough compromise might be to fetch +1 tree level collectively at once instead of 1 leaf at each time.

Reply:
@mvorisek Or shall I import the whole tree definition table into an array, and do the recursive operation inside the array? That would speed it up as well, and we talk about reasonably sized trees typically (otherwise we would need another treeview component which dynamically loads sub-trees upon click).

Reply:
It it obvious both extremes - current dummy impl. and whole table materialization is not a solution. The later case can produce OOM error easily, imagine the perf. on table with millions of records. It would be good to impl. recursive iterator in atk4/data Model and use it here.

Reply:
I think your worries are not realistic, as models for this TreeItemSelector are never tables of million entries. And if they are TreeItem is the wrong control as written before: you would need a Tree Control which dynamically loads sub-trees on clicks. As said, Recursive CTE is an overkill for this simple control, unless you see yourself in a position to implement this soon. Even with recursive CTE, the control would not load properly in time with your million-entries example. Otherwise I suggest we go for the faster export variant. Working well in real life with a few hundreds entries.

Reply:
Michael, let's focus on the solution and please reread my posts above. I did not written CTE is needed, I have written the current demo is awfully slow, which is a fact and I hope we both understand the PR cannot be merged as is.

I can also tell that table with millions of records is a realistic usecase, not for display, but imagine any tree features per some item like product. No product will have a lot of features, but the whole table can be huge. And as long as you cannot resolve the "minimal spanning tree" in one query, you cannot fetch the data using a single query.